upload/upload/main.py

```diff
@@ -0,0 +1,202 @@
+# useful links:
+# https://developers.google.com/api-client-library/python/
+# https://developers.google.com/api-client-library/python/apis/
+# https://developers.google.com/api-client-library/python/start/get_started
+# https://developers.google.com/identity/protocols/googlescopes
+
+import logging
+import re
+import datetime
+from flask import Flask, Response, session, request, redirect, render_template, escape, jsonify, abort, url_for
+import httplib2
+import json
+
+import google.oauth2.credentials
+import google_auth_oauthlib.flow
+
+from apiclient.discovery import build
+
+fmt = logging.Formatter(
+  # NB: no space after levename because WARNING is so long
+  '%(levelname)s\t| %(asctime)s \t| %(filename)s \t| %(funcName)s:%(lineno)d | '
+  '%(message)s')
+
+fh = logging.FileHandler('upload.log')
+fh.setLevel(logging.INFO)
+fh.setFormatter(fmt)
+
+ch = logging.StreamHandler()
+ch.setLevel(logging.INFO)
+ch.setFormatter(fmt)
+
+log = logging.getLogger('upload')
+log.setLevel(logging.INFO)
+
+logging.basicConfig(
+  handlers=[fh, ch],
+  level=logging.INFO)
+
+USERS = set([
+    'cseed@broadinstitute.org',
+    'dking@broadinstitute.org',
+    'jbloom@broadinstitute.org',
+    'jigold@broadinstitute.org',
+    'cvittal@broadinstitute.org',
+    'tpoterba@broadinstitute.org',
+    'pschultz@broadinstitute.org',
+    'wang@broadinstitute.org'
+])
+
+CLIENT_SECRET = '/upload-oauth2-client-secret/client_secret.json'
+SCOPES = [
+  'https://www.googleapis.com/auth/userinfo.email',
+  'https://www.googleapis.com/auth/userinfo.profile',
+  'https://www.googleapis.com/auth/plus.me'
+]
+
+counter = 0
+def next_id():
+    global counter
+
+    counter = counter + 1
+    return counter
+
+id_item = {}
+items = []
+
+class Item(object):
+    def __init__(self, data):
+        self.id = next_id()
+        
+        items.insert(0, self)
+        id_item[self.id] = self
+
+        self.timestamp = datetime.datetime.now()
+        self.data = data
+
+    def typ(self):
+        return self.data['type']
+
+    def email(self):
+        return self.data['email']
+
+    def config(self):
+        return self.data['config']
+
+    def hail_version(self):
+        return self.config()['hail_version']
+    
+    def contents(self):
+        return self.data['contents']
+
+    def preview(self):
+        s = self.contents()
+        if len(s) > 40:
+            s = s[:40] + " ..."
+        
+        return re.sub('\\s+', ' ', s)
+        
+        return s
+
+app = Flask('upload')
+
+with open('/flask-secret-key/flask-secret-key', 'rb') as f:
+    app.secret_key = f.read()
+
+def get_flow(state=None):
+  flow = google_auth_oauthlib.flow.Flow.from_client_secrets_file(
+    CLIENT_SECRET, scopes=SCOPES, state=state)
+
+  # FIXME _scheme='https'
+  flow.redirect_uri = url_for('oauth2callback', _external=True, _scheme='https')
+  
+  return flow
+
+@app.route('/', methods=['GET'])
+def index():
+    if 'email' not in session:
+        return redirect(url_for('login'))
```

Comment:
Ah, and this means the user logged in successfully modulo bugs in `upload` itself, because the `session` is signed by the secret flask key

Reply:
Yes.